### ISSUE TITLE
libjli.so: cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by ascherer** [processing/processing#5111]
Today, `processing` announced the new release 3.3.4. After downloading and extracting `processing-3.3.4.tgz` (both from the website and from github) for Linux 64-bit, and adjusting the
```
alias processing='CLASSPATH= ~/Work/tmp/processing-3.3.4/processing
```
I receive the Java error message
```
java: error while loading shared libraries: libjli.so: cannot open shared object file: No such file or directory
```
Curiously, `locate libjli.so` reports two entries:
```
/home/andreas/Work/tmp/processing-3.3.3/java/lib/amd64/jli/libjli.so
/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/jli/libjli.so
```
but `find ~/Work/tmp/processing-3.3.4 -name libjli.so` shows nothing. So it appears that Processing 3.3.4 is missing this dependency compared to 3.3.3. I'll revert to the previous version, which works fine.

**Comment by benfry** [comment]
If you use `./processing` from the `processing-3.3.4` directory, does it work? 

If you use `install.sh` does it work?

**Comment by jojomickymack** [comment]
looks like the linux 64 bit download for 3.3.4 is missing a bunch of shared object files in the java/lib/amd64 directory, also noted in the forum
https://forum.processing.org/two/discussion/22909/processing-3-3-4-won-t-start-on-linux

as a workaround I linked that directory to my jre, seems to work ok that way

**Comment by benfry** [comment]
Can you answer my two questions?

**Comment by linux-man** [comment]
> If you use ./processing from the processing-3.3.4 directory, does it work?

No
> If you use install.sh does it work?

install.sh does create the desktop file, but the error remains.

**Comment by ascherer** [comment]
afk

**Comment by ascherer** [comment]
Unpacking all Processing tarballs I've collected and used in the past without this issue reveals
```
andreas@komputer:~/Work/tmp$ find . -name libjli.so | sort
./processing-1.5.1/java/lib/i386/jli/libjli.so
./processing-2.0.3/java/lib/i386/jli/libjli.so
./processing-2.1/java/lib/i386/jli/libjli.so
./processing-2.2.1/java/lib/i386/jli/libjli.so
./processing-3.3.2/java/lib/amd64/jli/libjli.so
./processing-3.3.3/java/lib/amd64/jli/libjli.so
./processing-3.3/java/lib/amd64/jli/libjli.so
```
so it appears in fact to be version 3.3.4 that is missing this shared object.

**Comment by mxa** [comment]
IMO a broken version should be removed from the download section.

**Comment by ascherer** [comment]
Apart from the crappy message about “Java 8 from Oracle” against OpenJDK, I bring Processing 3.3.4 to life by replacing the whole `java/` subdirectory with a soft-link to `/usr/lib/jvm/java-8-openjdk-amd64/jre/` on Kubuntu 16.04. Maybe, just maybe, it would be a good idea to simply strip the bundled Java stuff (which expands to 82 MB on disk) from the tarball.

**Comment by monkstone** [comment]
@ascherer Unless you've already got it you would probably want to also install openjfx to support FX2D sketches, [JRubyArt](https://github.com/ruby-processing/JRubyArt) and [propane](https://github.com/ruby-processing/propane) work absolutely fine with openjdk.

**Comment by neilcsmith-net** [comment]
@ascherer +1 considering Oracle's JDK is 99% OpenJDK anyway, I do this too - be nice not to have the crappy message.  In fact, be nice if Processing became a fully open-source project and stopped using Oracle's JDK everywhere now that Azul are providing OpenJDK prebuilt for Windows and OSX.

**Comment by benfry** [comment]
The only thing that changed w/ the build processing in 3.3.4 on Linux is this: https://github.com/processing/processing/pull/5106
So it's possible we have a regression. Help wanted! 

If OpenJDK is 1% different from Oracle Java, then 1) it's the most important parts (how GUIs work, desktop integration, graphics handling), and 2) it shouldn't matter. Nobody's stepped up to consistently do Linux builds, so you're stuck with what I can support.

Bottom line, help me fix the issue at hand. This pile-on about OpenJDK is unproductive.

**Comment by mxa** [comment]
From what I can see it isn't related to https://github.com/processing/processing/pull/5106. There are many files missing from the java directory in the 3.3.4 version.
Size of java directory in the 3.3. download: 194.5 MB, in 3.3.4: 85.6 only.
Was it copied to the download before fully built?

**Comment by ascherer** [comment]
I haven't tried to build Processing locally yet. In case this is feasible, I think of writing a `processing.spec` file for packaging in rpm and deb formats with rpmbuild and debbuild. This would permit installation/deinstallation with respect to system dependencies and package relations. But don't hold your breath just now.

**Comment by neilcsmith-net** [comment]
@benfry I just took a zip clone of master and built on Ubuntu 16.04.  I should have taken a download of the release tag I guess, but I assume no relevant changes since?  Anyway, it builds and runs fine with default ant task.  I guess the only major difference may be I'm building with OpenJDK, but it''s downloading and running against Oracle JRE.  It doesn't look like there's any problems introduced with those build script changes.  Is the JRE downloaded during your build intact?

re. the above, I realise that piling on here is unproductive, so apologies for that, but please realise that it does matter for some people, so can we have a sensible discussion elsewhere?  Also, part of the issue here may be the hoops you have to go through to build against Oracle's JRE!  Not to mention that many people who might help out with Linux builds do so from a position of wanting to use open-source software.

**Comment by GKFX** [comment]
@neilcsmith-net The OpenJDK issue already exists at #4415. There's also a feature request at #4266 for Debian-packaged version that would automatically use the system Java, but due to other dependencies being unavailable that's not going to happen any time soon, unfortunately.

**Comment by neilcsmith-net** [comment]
@GKFX yes, I know, funnily enough (if you read it)! :smile:  The point was that GitHub issues are really bad for discussing overarching things.  @benfry mentioned a lack of people stepping up to help with Linux builds - I don't think those two things are unrelated!  But I don't want to derail this particular issue any further by it.

**Comment by monkstone** [comment]
@benfry this issue just seems to be a pile of missing files in the amd64 folder compared with previous processing releases.
This includes but possibly not limited to server/jvm.so and
libjsig.so
libzip.so
libverify.so
This is addition to missing jli/libjli.so

**Comment by benfry** [comment]
Thanks @monkstone. Not sure what happened there; will just have to do a release for 3.3.5 (to fix #5118 and #5110 as well) when I have the time.

**Comment by neilcsmith-net** [comment]
@benfry can you not just do a rebuild of the 3.3.4 tag?  As I mentioned already, this seems to be fine.  OTOH, if it doesn't build correctly for you it could be a sign of a problem on the build machine which would be good to pick up now?

**Comment by wolfgangwayland7** [comment]
Same as most claimed here. My-> OS : linux mint 18.1  
java: error while loading shared libraries: libjli.so: cannot open shared object file: No such file or directory
I'll have to back down one version...

**Comment by subham-s** [comment]
okay, you can copy the old java file and put it in new version. That will merge the other missing files (ofcourse not the license). that is working fine. glad i can make a fix!

**Comment by S6Regen** [comment]
The problem was flagged 14 days ago and you haven't been embarrassed enough to fix it yet?
I think the entire project has become a millstone around the original authors' necks.  Where they can't walk away from it, but they are as bored as heck by it.

**Comment by BrunoVernay** [comment]
Quit workaround:
`cd /your/path/processing-3.3.4`
`rm -rf java`
` ln -s /usr/lib/jvm/jre java`

**Comment by ShubhamPurandare** [comment]
Hey BrunoVernay ! Thank you for your answer it solved the error, though the processing started for a couple of seconds,  it has now lead to some other error named "Unsupported major.minor version 52.0"
Can anyone tell me what does this error mean ? Thank you in advance

**Comment by S6Regen** [comment]
I deleted the java folder inside processing and then it found the normal Java JDK I had on my system by itself!  It seems like such a simple problem to fix, what's the issue???  Perhaps it is some sort of legal trouble from Oracle?

**Comment by neilcsmith-net** [comment]
@ShubhamPurandare it means that the system JRE you're trying to link against is less than the required Java 8.  Is it OpenJDK 7 / Oracle Java 7?

**Comment by tim-taylor** [comment]
I have this problem too, on Kubuntu 17.04 64-bit. I've been using Processing for teaching for many years, and this is the first time that an update has failed to run out of the box. To have been running so well for so long across multiple platforms is a pretty impressive record by the team.

Bruno's workaround is problematic on K/ubuntu because Ubuntu uses IcedTea rather than Oracle Java 8, and Processing complains about this. I also needed to change the directory mentioned in Bruno's workaround to /usr/lib/jvm/default-java/jre. Having done this, when I run Processing I get a warning message that I'm not using the Oracle version of Java 8. Although the Processing window then opens, I have seen some crashes when trying to run stuff.

I tend to keep several older versions of Processing on the system, so rather than using the system version of Java, I've now switched to using the version supplied in the previous release, and this seems to work fine:

`cd /path/to/processing-3.3.4`
`mv java java.orig`
`ln -s /path/to/processing-3.3/java`

**Comment by neilcsmith-net** [comment]
@tim-taylor I make this "workaround" with every version of Processing anyway, and not yet seen crashes - might be useful to report those at #5135 (although is that meant to be IDE only @ascherer ?)

Incidentally, the other workaround is to build 3.3.4 from source yourself, which is pretty simple to do ( https://github.com/processing/processing/wiki/Build-Instructions ) and would also help prove if there is a build script problem.  As mentioned above, it builds fine for me.

**Comment by tim-taylor** [comment]
@neilcsmith-net TL;DR ignore my previous comment about crashes - I don't think it was Processing that was at fault.

I've now tested this further, and cannot reproduce the crash - I think it is more likely KDE that was at fault -- crashes seem to be not uncommon on it these days :-(

I've now run the problemmatic sketch (which is work submitted by a student, composed on a Mac) in Processing 3.3.3 (with its own problem-free copies of the Java libraries) on Kubuntu 17.04 both on my desktop and on my laptop. The program crashes with a runtime exception on my desktop, related to the ALSA sound system: "RtApiAlsa::probeDeviceOpen: pcm device (hw:0,3) won't open for input.
Could not run the sketch (Target VM failed to initialize)". On my laptop, it runs fine.  This is the same behaviour as when running Processing 3.3.4 and mentioned in my previous post.

The crash I referred to in my previous post happened after this runtime exception occurred on my desktop and I tried to close the Processing editor window - at that stage I got a system message saying that Processing had crashed. However, I can't reproduce that behaviour now - the window closes fine. As mentioned above, I do occasionally get crashes like this when trying to close other apps - so I think it is KDE to blame here.

**Comment by benfry** [comment]
Should be fixed for 3.3.5 which is now on the releases page. Still unsure what happened with the last round, but the new one is starting up just fine.

**Comment by mxa** [comment]
That's great. Thanks. But is it on https://processing.org/download/ ? I must be blind. Ctl+rrrrrrrrrrrrrrrr

**Comment by benfry** [comment]
I said the "releases page" (on this site: https://github.com/processing/processing/releases), you're looking at the download page. It takes a little longer to hit the download page since that has to be updated separately.
